### DISCUSSION TITLE
Why is being connected dependent on having data?

### DISCUSSION THREAD

**Discussion by marcelstoer** [Networking-for-Arduino/EthernetENC#31] [Q&A]
I have an MQTT [PubSubClient](https://github.com/knolleary/pubsubclient) over TLS+Ethernet using [SSLClient](https://github.com/OPEnSLab-OSU/SSLClient) and this library.

I am currently debugging weird warn/error messages from the SSLClient upon disconnecting from MQTT.

```
21:31:53.318 > (EthernetSSLClient)(SSL_ERROR)(m_update_engine): Error writing to m_client
21:31:53.323 > (EthernetSSLClient)(SSL_ERROR)(m_update_engine): 0
21:31:53.329 > (EthernetSSLClient)(SSL_WARN)(connected): Socket was dropped unexpectedly (this can be an alternative to closing the connection)
21:31:53.340 > (EthernetSSLClient)(SSL_ERROR)(connected): Not connected because write error is set
21:31:53.345 > (EthernetSSLClient)(SSL_ERROR)(m_print_ssl_error): SSL_CLIENT_WRITE_FAIL
21:31:53.351 > (EthernetSSLClient)(SSL_WARN)(m_run_until): Terminating with write error: 
21:31:53.357 > (EthernetSSLClient)(SSL_WARN)(m_run_until): 4
21:31:53.362 > (EthernetSSLClient)(SSL_ERROR)(flush): Could not flush write buffer!
21:31:53.368 > (EthernetSSLClient)(SSL_ERROR)(m_print_ssl_error): SSL_CLIENT_WRITE_FAIL
```

`PubSubClient#disconnect()` does this

```c++
void PubSubClient::disconnect() {
    this->buffer[0] = MQTTDISCONNECT;
    this->buffer[1] = 0;
    _client->write(this->buffer,2);
    _state = MQTT_DISCONNECTED;
    _client->flush(); // <------ THIS operation triggers the warnings downstream
    _client->stop();
    lastInActivity = lastOutActivity = millis();
}
```
Source: https://github.com/knolleary/pubsubclient/blob/master/src/PubSubClient.cpp#L660-L668

Upon flushing, the SSLClient then checks for the Ethernet client `getWriteError() || !connected()` in which case it prints this "Error writing to m_client" messages. I verified that `getWriteError()` returns 0 but....

This library's `connected()` confuses me:

https://github.com/jandrassy/EthernetENC/blob/a0393889ea90ea17e7ab8c33c759e1d6f16748e1/src/EthernetClient.cpp#L139

It assumes that if "no data, then disconnected", correct?

I confirmed that in my case `if (data) -> false`. However, I think it is expected that there is no data. The PubSub client consumed all MQTT messages by the time it initiates the disconnect. Hence, no data.

Maybe my reasoning above is flawed somewhere?

The same happens when reading data from an HTTPS connection using that SSLClient.

```c++
while(sslClient.connected()) {
  len = sslClient.available();
  if (len > 0) {
    // process
  }
}
sslClient.stop();
```
-> once we're done reading there's no more data available but since Ethernet client `connected()` checks whether there is data (and returns false) the SSLClient thinks we're disconnected

**Comment by JAndrassy** [comment]
Maybe one of these links will give you an answer:
[[https://www.arduino.cc/reference/en/libraries/ethernet/client.connected/](url)](https://www.arduino.cc/reference/en/libraries/ethernet/client.connected/)
https://github.com/OPEnSLab-OSU/SSLClient/issues/43

**Reply by marcelstoer** [reply]
Thanks! Those are very helpful references.
